internal/gengapic/paging.go

```diff
@@ -67,8 +57,33 @@ func (g *generator) iterTypeOf(elemField *descriptor.FieldDescriptorProto) (*ite
 			typ = parent
 		}
 
-		pt.elemTypeName = fmt.Sprintf("*%s.%s", imp.Name, typeName)
-		pt.iterTypeName = typeName + "Iterator"
+		eMsg, ok := eType.(*descriptor.DescriptorProto)
+		if !ok {
+			return nil, errors.E(nil, "cannot find message type %q, malformed descriptor", eType)
+		}
+		if eMsg.GetOptions().GetMapEntry() {
+			var valueField *descriptor.FieldDescriptorProto
+			for _, f := range eMsg.GetField() {
+				if f.GetName() == "value" {
+					valueField = f
+					break
+				}
+			}
+			if valueField == nil {
+				return nil, errors.E(nil, "unusual map entry message: %q", eMsg)
+			}
+
+			if valueField.GetType() == descriptor.FieldDescriptorProto_TYPE_MESSAGE {
+				pt.mapValueTypeName = valueField.GetTypeName()
+			} else {
+				pt.mapValueTypeName = pbinfo.GoTypeForPrim[valueField.GetType()]
+			}
+			pt.elemTypeName = fmt.Sprintf("String%sPair", upperFirst(pt.mapValueTypeName))
+			pt.iterTypeName = pt.elemTypeName + "Iterator"
+		} else {
+			pt.elemTypeName = fmt.Sprintf("*%s.%s", imp.Name, typeName)
+			pt.iterTypeName = typeName + "Iterator"
+		}
```

Comment:
Nit, style: Can you move this out of the `else` to just above the `if`? I prefer to do the base case (if it has no side effects by doing so) by default, then handle special cases with `if` following it. This keep sthe main flow of the program left-justified.

Reply:
Done